Implement support for DWARF version 5.

pcwalton opened the pull request:
DWARF version 5 brings a number of improvements over version 4. Quoting from
the announcement [1]:

> Version 5 incorporates improvements in many areas: better data compression,
> separation of debugging data from executable files, improved description of
> macros and source files, faster searching for symbols, improved debugging
> optimized code, as well as numerous improvements in functionality and
> performance.

On platforms where DWARF version 5 is supported (Linux, primarily), this commit
adds support for it behind a new `-Z dwarf-version=5` flag.

[1]: https://dwarfstd.org/Public_Review.php

r? @michaelwoerister

nikic commented:
This should probably be an unstable `-Z dwarf-version` flag, unless you intend this to be insta-stable.

Is the max dwarf version logic really needed? It seems odd that there is no way to force a newer version if a target has this option set. It's reasonable that some platforms default to an older version, but I don't really see a reason to hard forbid using a newer one (for example, if I bring my own toolchain).

pcwalton commented:
OK, I've switched the `-C` flag to a `-Z` flag, removed the `max_dwarf_version` field from the target info, and moved the documentation over to the unstable book.

nnethercote commented:
Is this worth a perf run?

nikic commented on src/test/assembly/dwarf5.rs:
```diff
@@ -0,0 +1,20 @@
+// Makes sure that `-C dwarf-version=5` causes `rustc` to emit DWARF version 5.
```
`-C` -> `-Z` in this comment.

nikic reviewed (commented):
Looks reasonable to me.

michaelwoerister commented:
Thanks, @pcwalton! Looks good to me.

@bors r+

> Is this worth a perf run?

As it currently is, the PR does not actually change the DWARF version being used for anything. It just adds the `-Z` flag for explicitly setting it. So, we shouldn't see a perf difference. But once this is merged it would be interesting to see how switching to v5 affects performance.

bors commented:
:pushpin: Commit 6c73840877fb5db8fd84c431d7078abe23cd8f87 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 6c73840877fb5db8fd84c431d7078abe23cd8f87 -->
<!-- homu: {"type":"Approved","sha":"6c73840877fb5db8fd84c431d7078abe23cd8f87","approver":"michaelwoerister"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/98381#issuecomment-1163241440) for pull requests below priority 1000. This pull request will be tested once the tree is reopened.

nikic commented:
@bors r=michaelwoerister

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7010525213?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/42eeb58e5dc8d409d0f1db40486f9f8db4436913/checks/7010525213/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

pcwalton commented:
Doesn't look related to this PR.

nnethercote commented:
That failure is the one that caused the tree closure. Should be resolved now.

@bors retry

michaelwoerister commented:
@bors retry

michaelwoerister commented:
@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #98566

bors commented:
:pushpin: Commit 42eeb58e5dc8d409d0f1db40486f9f8db4436913 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 42eeb58e5dc8d409d0f1db40486f9f8db4436913 -->
<!-- homu: {"type":"Approved","sha":"42eeb58e5dc8d409d0f1db40486f9f8db4436913","approver":"michaelwoerister"} -->

Dylan-DPC commented:
failed in [rollup](https://github.com/rust-lang-ci/rust/runs/7108528451?check_suite_focus=true)

@bors r-

pcwalton commented:
@bors retry

Looks like that failing test got reworked to not depend on -Z flags.

bors commented:
@pcwalton: :key: Insufficient privileges: not in try users

pcwalton commented:
Could anyone retry this? I believe the test is fixed now.

dtolnay commented:
@bors retry

ehuss commented:
@bors r=michaelwoerister

retry does not work when it has been unaccepted.

bors commented:
:pushpin: Commit 1e0ad0c1d4fdca7a37221e05fd5e39739d9084f5 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister 1e0ad0c1d4fdca7a37221e05fd5e39739d9084f5 -->
<!-- homu: {"type":"Approved","sha":"1e0ad0c1d4fdca7a37221e05fd5e39739d9084f5","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

eddyb commented on compiler/rustc_codegen_llvm/src/debuginfo/mod.rs:
```diff
@@ -103,14 +103,14 @@ impl<'ll, 'tcx> CodegenUnitDebugContext<'ll, 'tcx> {
-                    version,
-                )
-            }
+            let dwarf_version =
+                sess.opts.debugging_opts.dwarf_version.unwrap_or(sess.target.default_dwarf_version);
+            llvm::LLVMRustAddModuleFlag(
+                self.llmod,
+                llvm::LLVMModFlagBehavior::Warning,
+                "Dwarf Version\0".as_ptr().cast(),
+                dwarf_version,
+            );
```
Could this have caused #99143? Looks like before this PR, we only ever set `Dwarf Version` on targets with `Some` in their `dwarf_version`, which based on the diff in this PR, wasn't the case for e.g. Windows (at least, there's probably more targets not covered by the `${OS}_base.rs` files).

eddyb replied:
Alright that was confusing but I'm pretty sure (https://github.com/rust-lang/rust/issues/99143#issuecomment-1182315377) that what's going on is even with debuginfo disabled for a binary, the standard library is getting statically linked in *and we ship it w/ debuginfo*.

I'm not sure what the best approach here would be, maybe just relying on the `is_like_msvc` check to effectively switch between `CodeView` and `Dwarf Version` (and never set both)?

Wyvern commented:
Does macOS support DWARF5 at the moment?
